# Python Data Types

In [2]:
print(str, int, float, bool)

<class 'str'> <class 'int'> <class 'float'> <class 'bool'>


In [4]:
type(3)

int

In [46]:
type(3.0)

float

#### They're different types, different representation
But Python implicitly converts `3` to `float` to compare "apples to apples".

In [6]:
3.0 == 3

True

Same, but this time we explicitly tell Python to perform the conversion

In [9]:
3.0 == float(3)

True

Implicit conversion doesn't happen when comparing to `str`

In [7]:
'3' == 3

False

But we can do it explicitly

In [10]:
'3' == str(3)

True

In [11]:
int('3') == 3

True

One final primitive type is `None`. It literally means "nothing".

In [14]:
None

In [15]:
type(None)

NoneType

### Operations
Basic arithmetic.

In [19]:
2+4*3/8

3.5

Same, but without the division, and we get `int` as a result

In [20]:
2+4*3

14

In [21]:
5 / 3

1.6666666666666667

5 divided by 3 is 1.667, and this takes the integer part

In [22]:
5 // 3

1

8 divided by 3 is 2, 8 - 2*3 = 2. This 2 remains after division

In [47]:
8 % 3

2

In [24]:
7 % 2

1

In [25]:
8 % 2

0

In [26]:
8/3

2.6666666666666665

In [27]:
8 // 3

2

In [28]:
8 % 3

2

#### Powers
3 to the power of 4 means `3*3*3*3` - multiply 3 a total of 4 times.

In [31]:
3**4

81

In [32]:
3*3*3*3

81

The powers of 2.

|power|multiplication|value|
|----:|:-------------|----:|
|0|`1`|1|
|1|`1*2`|2|
|2|`1*2*2`|4|
|3|`1*2*2*2`|8|
|4|`1*2*2*2*2`|16|
|5|`1*2*2*2*2*2`|32|
|6|`1*2*2*2*2*2*2`|64|
|7|`1*2*2*2*2*2*2*2`|128|
|8|`1*2*2*2*2*2*2*2*2`|256|

In [48]:
2**8

256

The power (called *"exponent"*) can also be a non-integer

In [35]:
256**0.1236234

1.9847911396630227

The same operation can mean different things depending on the `type` of the values you perform it on

In [36]:
"Hello, " + "World!"

'Hello, World!'

##### Why are powers useful:
instead of writing this and counting 9 zeroes

In [37]:
1000000000

1000000000

You can choose to just raise 10 to the power of 9, so you know you have 9 zeroes

In [38]:
10**9

1000000000

But it's **especially** useful for the powers of 2

*though below I used latin names for `kilo`, `mega`, `giga`, they actually mean exact thousands, millions, etc. New names were invented for the computer-based "almost" numbers, called instead `kibibyte`, `mebibyte`, `gibibyte`, etc. but people mostly talk in `kilo` still*

Even though you can remember 1024 as 2 to the power of 10

In [39]:
#1 kilobyte = 1024 bytes
2**10 # 2*2*2*2*2*2*2*2*2*2

1024

... most people will not bother with remembering 2 to the power of 20 and beyond

In [41]:
#1 megabyte = 1024 kilobytes = 1024 * 1024 bytes
2**20 # 2*2*2*2*2*2*2*2*2*2 * 2*2*2*2*2*2*2*2*2*2

1048576

In [42]:
#1 gigabyte = 1024 megabytes = 1024 * 1024 * 1024 bytes
# 2*2*2*2*2*2*2*2*2*2 * 2*2*2*2*2*2*2*2*2*2 * 2*2*2*2*2*2*2*2*2*2
2**30

1073741824

In [43]:
#1 terabyte = 1024 gigabytes = 1024 * 1024 * 1024 * 1024 bytes
2**40

1099511627776

In [44]:
#1 petabytes = 1024 terabyte = 1024 * 1024 * 1024 * 1024 * 1024 bytes
2**50

1125899906842624